Change ListViewCard observables to PublishSubjects

CoreyHorn opened the pull request:
These changes allow a consumer of the API to subscribe to the Observables surfaced by ListViewCard at any point. Previous implementation could lead to NPEs being thrown when attempting to subscribe too early.

CoreyHorn committed:
Change ListViewCard observables to PublishSubjects

-Up gradle version

mdesoto commented:
I don't think this is the right approach. We have naked subscriptions that cannot be unsubscribed from, and we're creating one for each item. Which isn't necessary either.

mdesoto commented on library/src/main/java/io/andref/rx/widgets/ListViewCard.java:
```diff
@@ -106,7 +107,8 @@ private void initializeViews(@NonNull Context context, @Nullable AttributeSet at
         mButton.setText(buttonText);
-        mButtonClicks = RxView.clicks(frameLayout);
+        RxView.clicks(frameLayout)
+                .subscribe(mButtonClicks);
```
Don't like the naked subscriptions buried in the view. Need to make sure they get unsubscribed from. Maybe `takeUntil`, `RxView.detaches()`. Or something like that.

CoreyHorn replied:
Good point. Doing that now.

mdesoto commented on library/src/main/java/io/andref/rx/widgets/ListViewCard.java:
```diff
@@ -183,8 +182,18 @@ private void layoutViews()
+                            return item;
+                        }
+                    })
+                    .subscribe(mItemClicks);
```
Same as above. Plus we're creating a subscription for every view.  Seems unnecessary.

CoreyHorn replied:
If we want to not have subscriptions for every view, we will need to switch to using the adapter pattern for this view. Any work we do to support managing multiple views within this one would just be duplicating that pattern. While I think we will want to do that eventually, I don't think this is the time and place.

mdesoto replied:
You should be able to merge all the observables and subscribe to that.

mdesoto commented on library/src/main/java/io/andref/rx/widgets/ListViewCard.java:
```diff
@@ -193,26 +202,11 @@ public Item call(Void aVoid)
-                                    }
-                                })
-                );
+                        .subscribe(mIconClicks);
```
Same as above (x2).

CoreyHorn committed:
Manage Subscriptions Properly

-Add CompositeSubscription to ListViewCard and
unsubscribe when needed.

mdesoto commented on library/src/main/java/io/andref/rx/widgets/ListViewCard.java:
```diff
@@ -76,6 +79,20 @@ public ListViewCard(@NonNull Context context, @Nullable AttributeSet attrs, @Att
+    {
+        mCompositeSubscription.unsubscribe();
+        super.onDetachedFromWindow();
+    }
```
Still don't think we have the right answer here. We need to drive/respond to this view in the consuming activity/fragment, and that lifecycle won't necessarily match window state.

If we're going to subscribe in the view, I think `takeUntil(view detaches)` is more appropriate.

CoreyHorn committed:
Code Review Changes

-Merge all observables and surface as one
-Ensure all subscriptions are unsubscribed properly

mdesoto committed:
Reverted back to original approach to merging item related observables. Kept publish subjects.

mdesoto committed:
Added ability to add/remove items and to show/hide button. These things are now testable in the example app.

mdesoto commented:
OK! I refactored this a bit. And added the ability to manipulate the view (show/hide button, add/remove items). Frankly those should be UI tests.

Anyway, have a look. See if there's anything I missed.

CoreyHorn commented on library/src/main/java/io/andref/rx/widgets/ListViewCard.java:
```diff
@@ -29,23 +29,25 @@
+    private final PublishSubject<Item> mItemClicks = PublishSubject.create();
+    private final PublishSubject<Item> mIconClicks = PublishSubject.create();
+
     private Observable<Void> mButtonClicks;
```
Any reason you converted this back to a plain Observable as opposed to a PublishSubject? 

Using a PublishSubject allows subscribing at any point and seems to be good practice for surfacing view bindings as observables.

CoreyHorn replied:
After removing all views and adding another, none of the clicks on the newly added views are being recognized

mdesoto replied:
The button is never removed, so there's no need to wrap it's observable in a publish subject.

CoreyHorn replied:
The problem is that we are exposing a null Observable. Attempting to subscribe or call operators on that Observable will cause NPEs if it hasn't been initialized yet. I usually fix that by exposing PublishSubjects. 

However, we may also be able to fix that by initializing the Observables as Observable.empty().

CoreyHorn committed:
Initialze Empty Observable
